server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -124,19 +124,31 @@ protected void processInput(DesiredBalanceInput desiredBalanceInput) {
                 long index = desiredBalanceInput.index();
                 logger.debug("Starting desired balance computation for [{}]", index);
 
-                recordTime(
-                    cumulativeComputationTime,
-                    () -> setCurrentDesiredBalance(
-                        desiredBalanceComputer.compute(
-                            getInitialDesiredBalance(),
-                            desiredBalanceInput,
-                            pendingDesiredBalanceMoves,
-                            this::isFresh
-                        )
-                    )
-                );
+                DesiredBalance desiredBalance;
+                final long started = threadPool.relativeTimeInMillis();
+                try {
+                    desiredBalance = desiredBalanceComputer.compute(
+                        getInitialDesiredBalance(),
+                        desiredBalanceInput,
+                        pendingDesiredBalanceMoves,
+                        this::isFresh
+                    );
+                    setCurrentDesiredBalance(desiredBalance);
+                } finally {
+                    final long finished = threadPool.relativeTimeInMillis();
+                    cumulativeComputationTime.inc(finished - started);
+                }
                 computationsExecuted.inc();
-                if (isFresh(desiredBalanceInput)) {
+
+                if (desiredBalance.finishReason() == DesiredBalance.ComputationFinishReason.STOP_EARLY) {
+                    logger.debug(
+                        "Desired balance computation for [{}] terminated early with partial result, scheduling reconciliation",
+                        index
+                    );
+                    submitReconcileTask(currentDesiredBalance);
+                    var newInput = DesiredBalanceInput.create(indexGenerator.incrementAndGet(), desiredBalanceInput.routingAllocation());
+                    desiredBalanceComputation.compareAndEnqueue(desiredBalanceInput, newInput);
```

Comment:
It is a bit confusing that condition uses `desiredBalance` but the branch submits `currentDesiredBalance`.
It took me a minute to understand they are pointing to the same object. 
It would be nice to replace them with a single ref to avoid this confusion.

Reply:
this shouldn't be an issue since I addressed your previous comment.